add docs for smoke testing of sso authentication (#838)
nextcloud/integration_openproject: docs/testing/smoke_testing.md

- [Section A: OAuth 2.0](#section-a-oauth-20-two-way-authorization-code-flow)
- [Section B: SSO through OpenID Connect](#section-b-single-sign-on-sso-through-openid-connect)
  - [Section B.1: Nextcloud Hub](#section-B1-nextcloud-hub)
  - [Section B.2: External Provider](#section-B2-External-Provider)

- [Common Smoke Test Steps](#common-smoke-test-steps)
  - [1. Link/Unlink a work package for a file/folder in Nextcloud](#1-linkunlink-a-work-package-for-a-filefolder-in-nextcloud)  
  - [2. Link/Unlink a work package for a file/folder from OpenProject](#2-linkunlink-a-work-package-for-a-filefolder-from-openproject)  
  - [3. Direct upload file/folder from OpenProject to Nextcloud](#3-direct-upload-filefolder-from-openproject-to-nextcloud)  
  - [4. Create a WorkPackage from Nextcloud](#4-create-a-workpackage-from-nextcloud)  
  - [5. Check notification in `OpenProject` widget in Nextcloud](#5-check-notification-in-openproject-widget-in-nextcloud)  
  - [6. Check New folder with automatically managed permissions in OpenProject](#6-check-new-folder-with-automatically-managed-permissions-in-openproject)  

## Section A: OAuth 2.0 (Two-Way Authorization Code Flow)

> ### A1. Oauth configuration (without project folder setup/Automatically managed folders)

Reviewer: ```suggestion
### A1. OAuth configuration
```

- [ ] In `OpenProject`, navigate to `Administration > Files`.
- [ ] Create a file storage type `Nextcloud` and name it as `Nextcloud` in `OpenProject`.
- [ ] In admin setting of `Nextcloud`, navigate to `Administration Settings > OpenProject`.
- [ ] Copy `OpenProject` Oauth Credential (client_id and client_secret) and save them in `Nextcloud`.

> - [ ] Copy `Nextcloud` Oauth Credential (client_id and client_secret) and save them in `OpenProject`.

Reviewer: ```suggestion
- [ ] Copy `Nextcloud` Oauth Credential (client_id and client_secret) and save them in `OpenProject`.
- [ ] Disable project folder (automatically managed folder).
```

> ### A2. Connect Nextcloud with OpenProject (Without project folder setup)

Reviewer: ```suggestion
### A2. Connect Nextcloud with OpenProject
```

> - [ ] Complete Smoke [Test No A1](#a1-oauth-configuration-without-project-folder-setupautomatically-managed-folders).

Reviewer: ```suggestion
- [ ] Complete step [Test No A1](#a1-oauth-configuration-without-project-folder-setupautomatically-managed-folders).
```

- [ ] Log out, then Login as admin in `Openproject`.
- [ ] As an `OpenProject` admin, add keycloak-created user as a member in one of the project.

### B10. Configure integration_openproject with keycloak (token exchange disable)
- [ ] In nextcloud, go to `Administration > OpenProject`.
- [ ] Under `Authentication Method`, select `Single-Sign-On through OpenID Connect Identity Provider`.
- [ ] In `Authentication settings`, select `provider Type` as `Keycloak`.
- [ ] Disable `token exchange`

### B11. Verify Connection in nextcloud
- [ ] Complete step [Test No B7](#b7-Configure-Keycloak-IdP-Setup).
- [ ] Complete step [Test No B8](#b8-Configure-Nextcloud-Client).
- [ ] Complete step [Test No B9](#b9-Configure-OpenProject-Client).
- [ ] Complete step [Test No B10](#b10-Configure-integration_openproject-with-keycloak-token-exchange-disable).
- [ ] In nextcloud, login as keycloak-created user.
- [ ] Navigate to `settings > Openproject`
- [ ] Should show user is connected as an OpenProject user.

### B12. Complete the 6 common tests
- [ ] Complete steps [Test No [1-6]](#common-smoke-test-steps-applies-to-both-authentication-methods).

> ### B13. Configure integration_openproject with keycloak (token exchange enable)

Reviewer: ```suggestion
### B13. Setup integration (token exchange enabled)
```

- [ ] Complete step [Test No B7](#b7-Configure-Keycloak-IdP-Setup).
- [ ] Complete step [Test No B8](#b8-Configure-Nextcloud-Client).
- [ ] Complete step [Test No B9](#b9-Configure-OpenProject-Client).
- [ ] Go to `Administration > OpenProject` in nextcloud
- [ ] Under `Authentication Method`, select `Single-Sign-On through OpenID Connect Identity Provider`
- [ ] In `Authentication settings`, select `provider Type` as `Keycloak`
- [ ] Enable `token exchange`
- [ ] Set `OpenProject client ID *` as `Openproject`

### B14. Verify Connection in nextcloud
- [ ] Complete step [Test No B7](#b7-Configure-Keycloak-IdP-Setup).
- [ ] Complete step [Test No B8](#b8-Configure-Nextcloud-Client).
- [ ] Complete step [Test No B9](#b9-Configure-OpenProject-Client).
- [ ] Complete step [Test No B11](#b11-Verify-Connection-in-nextcloud).
- [ ] In nextcloud, login as keycloak-created user.
- [ ] Navigate to `settings > Openproject`
- [ ] Should show user is connected as an OpenProject user.

### B15. Complete the 6 common tests
- [ ] Complete steps [Test No [1-6]](#common-smoke-test-steps-applies-to-both-authentication-methods).

> ### B16. Check the integration script for sso setup (idp as nextcloud)

Reviewer: ```suggestion
### B16. Check the integration script for sso setup (Nextcloud Hub)
```

> Before Running the script make sure that your `Nextcloud` and `OpenProject` instance is up and running
> If you're using Nextcloud as the Identity Provider (OIDC), make sure the following apps are installed and enabled in Nextcloud:
>   - oidc
>   - integration_openproject
>
> If you are using a custom Identity Provider, ensure that:
>   - user_oidc app is installed and enabled in Nextcloud.
>   - The custom IdP is properly configured and accessible.
>
> To add the Nextcloud storage, delete the 'nextcloud' file storage from OpenProject, reset the Nextcloud config, and run the script again.

> ### Check to set up sso where nextcloud as IdP

Reviewer: ```suggestion
```

```bash
NC_INTEGRATION_PROVIDER_TYPE=nextcloud_hub \
NC_ADMIN_USERNAME=admin \
NC_ADMIN_PASSWORD=admin \
NC_INTEGRATION_ENABLE_NAVIGATION=false \
NC_INTEGRATION_ENABLE_SEARCH=false \
NC_HOST=https://nextcloud.local \
OP_ADMIN_USERNAME=admin \
OP_ADMIN_PASSWORD=admin \
OP_STORAGE_NAME=nextcloud \
OP_HOST=https://openproject.local \
OP_USE_LOGIN_TOKEN=true \
bash integration_oidc_setup.sh
```

- [ ] Upon success, try step [Test No B4](#b4-Use-SSO-button-of-nextcloud-in-OpenProject-for-login).
- [ ] Upon success, try step [Test No B5](#b5-Verify-Connection-of-Nextcloud-user-with-OpenProject).
- [ ] Also, to set up the integration configuration with project folder setup, at first delete 'nextcloud' file storage from OpenProject.
- [ ] In nextcloud, delete the `OpenProject` user, group and team folder from the nextcloud (if they exist).
- [ ] Then, reset the Nextcloud config.
- [ ] set environment `SETUP_PROJECT_FOLDER=true` and run the script.
- [ ] Run the script again after it is already setup (Should not give any error).

### Check to set up sso where keycloak as IdP (token exchange disable)

> > Before Running the below script make sure that you delete the 'nextcloud' file storage from OpenProject and reset the Nextcloud config.

Reviewer: ```suggestion
> Before running the script make sure that you delete the 'nextcloud' file storage from OpenProject and reset the integration settings in Nextcloud.
```

- [ ] Run the `integration_oidc_setup.sh` script to set up integration without project folder with the following command:

```bash
NC_HOST=https://nextcloud.local \
NC_ADMIN_USERNAME=admin \
NC_ADMIN_PASSWORD=admin \
NC_INTEGRATION_PROVIDER_TYPE=external \
NC_INTEGRATION_PROVIDER_NAME=keycloak \
NC_INTEGRATION_OP_CLIENT_ID=openproject \
NC_INTEGRATION_TOKEN_EXCHANGE=false \
NC_INTEGRATION_ENABLE_NAVIGATION=false \
NC_INTEGRATION_ENABLE_SEARCH=false \
OP_HOST=https://openproject.local \
OP_ADMIN_USERNAME=admin \
OP_ADMIN_PASSWORD=admin \
OP_STORAGE_NAME=nextcloud \
OP_USE_LOGIN_TOKEN=true \
bash integration_oidc_setup.sh

```

- [ ] Upon success, try step [Test No B11](#B11-Verify-Connection-in-nextcloud).
- [ ] Also, to set up the integration configuration with project folder setup, at first delete 'nextcloud' file storage from OpenProject.
- [ ] In nextcloud, delete the `OpenProject` user, group and team folder from the nextcloud (if they exist).
- [ ] Then, reset the Nextcloud config.
- [ ] set environment `SETUP_PROJECT_FOLDER=true` and run the script.
- [ ] Run the script again after it is already setup (Should not give any error).

> ### Check to set up sso where keycloak as IdP (token exchange enable)

Reviewer: ```suggestion
### Check the integration script for sso setup (External provider  with token exchange)
```
